### PR TITLE
GH-38460: [Java][FlightRPC] Add mTLS support for Flight SQL JDBC driver

### LINKED ISSUE #38460: [Java][FlightRPC] Add mTLS support to the Arrow Flight SQL JDBC driver
### Describe the enhancement requested

This is to request the addition of mTLS client verification/authentication to the Arrow Flight SQL JDBC driver.

Currently - Arrow Flight SQL ADBC has support for this, but the JDBC driver does not.  This would bring parity between the drivers.

### Component(s)

FlightRPC

### PR DESCRIPTION
**Pull request by prmoore77**
### Rationale for this change

I wanted to add additional security capabilities to the Arrow Flight SQL JDBC driver so that it catches up to ADBC.  ADBC already supports mTLS - and it is a great security feature.  I wanted to bring this to the JDBC driver as well.

### What changes are included in this PR?

This PR adds support for mTLS (client certificate verification/authentication) to the Arrow Flight SQL JDBC driver.

### Are these changes tested?

Yes, I've added tests of the new mTLS functionality - and have ensured that the change is backward compatible by verifying all existing tests pass.

### Are there any user-facing changes?

Yes - but the end-user documentation for the Arrow Flight SQL JDBC driver has been updated in the PR itself.

* Closes: apache/arrow#38460
* GitHub Issue: #38460

### CODE REVIEW

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (jduo) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:286]: try-with-resources instead

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (jduo) [inline docs/source/java/flight_sql_jdbc_driver.rst:122]: It's a bit odd that these new parameters are raw PEM files instead of JKS files like the other TLS-related parameters.

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (jduo) [inline docs/source/java/flight_sql_jdbc_driver.rst:129]: Maybe clientPrivateKey would be more clear?

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (prmoore77) [reply docs/source/java/flight_sql_jdbc_driver.rst:122]: Hi @jduo - I was trying to add a convenience factor.  For me, it is easier to work with PEM files - I added tlsRootCerts for folks that prefer PEM.  I had been using PEM files with ADBC already - and it seemed so convenient, I wanted the option for JDBC...

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (prmoore77) [reply docs/source/java/flight_sql_jdbc_driver.rst:129]: agreed - I'll make the change...

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:286]: Thanks for the tip!  I'll make the change...

**Comment by jbonofre** [comment]
Thanks for the PR. I'm doing a pass and testing as well.

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (prmoore77) [reply docs/source/java/flight_sql_jdbc_driver.rst:129]: Oh - I realized I used `clientKey` because the Java Flight client has it named that way:
https://github.com/apache/arrow/blob/57f643c2cecca729109daae18c7a64f3a37e76e4/java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightClient.java#L661C1-L661C1

Are you ok with the name (for consistency's sake)?

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (danepitkin) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:415]: We could use `@ParameterizedTest` to reduce some code duplication in the test files.

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (danepitkin) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:234]: Out of curiosity, why is there a 300 second timeout for ArrowFlightJdbcDataSource connections, but 0 second timeouts for `ArrowFlightJdbcDriver`/`DriverManager` connections?

**Review by danepitkin — COMMENTED** [review]
The tests are quite thorough, nice work! I'll defer to others who have more experience on Flight SQL for approval.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:348]: If the user calls this method repeatedly we will leak File handles. Need to close the current one if it exists.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:368]: Similar comment as above about this potentially leaking a file. Also, now it is ambiguous whether the Flight client should close() mTlsCACert (presumably it should).

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:234]: Very good question :) - I copied the tests from the `ConnectionTlsTest.java` code - while modifying the tests for mTLS of course.  I'm not sure of the rationale for the timeout, but I'll research it.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:348]: Hey @jduo - I'm not a very good Java developer - but I'm learning (thanks for your patience).  Would it be like this?
```java
    public Builder useMTlsClientVerification(final File mTlsCACert) throws IOException {
      if (this.mTlsCACert != null) {
            this.mTlsCACert.close();
        }
      this.mTlsCACert = new FileInputStream(mTlsCACert); 
```
Thanks for your help!

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:368]: I mirrored the existing `useTls` method [here](https://github.com/apache/arrow/blob/fcbcb7dab71b14ff3436af7b21bd23132877e2fa/java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java#L315)

Does that code have the issue as well?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:348]: Yes this would work. However I'm leaning towards deferring opening mTlsCACert until the connection is made. It's odd for builders to have more logic than setting fields.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:368]: Correct. It looks to have this issue too.

**Review by davisusanibar — COMMENTED** [review]
Thank you for the PR. I am adding a few minor questions.

**Comment by prmoore77** [comment]
Hello – it could be added, but the Java Flight client doesn’t yet support that to my knowledge – so I just wanted to catch up to that for now.  We can add it in a later PR, unless you feel strongly that it should be included here.

From: david dali susanibar arce ***@***.***>
Date: Saturday, October 28, 2023 at 7:20 PM
To: apache/arrow ***@***.***>
Cc: prmoore77 ***@***.***>, Mention ***@***.***>
Subject: Re: [apache/arrow] GH-38460: [Java][FlightRPC] Add mTLS support for Flight SQL JDBC driver (PR #38461)

@davisusanibar commented on this pull request.

Thank you for the PR. I am adding a few minor questions.

________________________________

In docs/source/java/flight_sql_jdbc_driver.rst<https://github.com/apache/arrow/pull/38461#discussion_r1375164054>:

> @@ -114,6 +114,21 @@ parameters are:

      - null

      - When TLS is enabled, the password for the certificate store



+   * - tlsRootCerts

+     - null

+     - Path to PEM-encoded root certificates for TLS - use this as

+       an alternative to ``trustStore``

+

+   * - clientCertificate

+     - null

+     - Path to PEM-encoded client mTLS certificate when the Flight

+       SQL server requires client verification.

+

+   * - clientKey

Will it be necessary to provide an optional client key password argument if this client private key is password-protected?

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java
Comment (vibhatha) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java:470]: nit: shall we add documentation?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (vibhatha) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:348]: @jduo just for my knowledge and to clarify a point, would it be more easier if we use try-with-resources here?

**Comment by vibhatha** [comment]
@prmoore77 I left a few comments and nice work with the test cases and comments. Thanks for working on this issue.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:362]: hi @jduo - this is the FlightServer - I'd rather not change the overall design of the current class's builder.  It already uses this approach for `useTls` - and I think my change is consistent with that.

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java:476]: hi, @vibhatha - Noted - corrected and pushed a change just now.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:385]: @vibhatha @davisusanibar can you file an issue to fix this?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:362]: @vibhatha , try-with-resources wouldn't help here since we need the stream to have a longer lifetime than this builder function. This is more of a case of tracking ownership.

Comment (vibhatha) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:362]: Got it, thanks @jduo

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (vibhatha) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:385]: @lidavidm I will file one. Will study a bit and file one.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:362]: @jduo - I've added small changes that detect if the InputStream attributes are not null, and if so - closes them.  Could you please re-review?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:385]: given that this issue will be filed separately - is it acceptable to approve/merge this PR?

File: docs/source/java/flight_sql_jdbc_driver.rst
Comment (prmoore77) [reply docs/source/java/flight_sql_jdbc_driver.rst:127]: Hello – it could be added, but the Java Flight client doesn’t yet support that to my knowledge – so I just wanted to catch up to that for now.  We can add it in a later PR, unless you feel strongly that it should be included here.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:350]: We should set these to null after closing, because there can be an exception opening the new inputs. So this would leave these fields in a bad state (holding closed streams).

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:362]: Thanks @prmoore77 minor comment about another possible error scenario in useTls() that applies here.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:269]: https://netty.io/4.1/api/io/netty/handler/ssl/SslContextBuilder.html#trustManager-java.io.InputStream-

```suggestion
        try {
          sslContext = sslContextBuilder.build();
        } catch (SSLException e) {
          throw new RuntimeException(e);
        } finally {
          if (mTlsCACert != null) {
            mTlsCACert.close();
          }
          certChain.close();
          key.close();
        }
```

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java:252]: Why the redundant checkNotNull?

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionTlsRootCertsTest.java
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionTlsRootCertsTest.java:120]: Use assertThrows

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java:139]: Use assertThrows

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:356]: would it be like this?:
```
    private void closeInputStreamIfNotNull(InputStream stream) throws IOException {
      if (stream != null) {
        stream.close();
      }
      stream = null;
    }
```

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java:252]: Good point.  I copied the approach from: https://github.com/apache/arrow/blob/cd6e63570f81e96375a4c51ef5d925b5f32f5a57/java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java#L225 - should we remove there also?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:356]: No, because that only sets the local - you need to set `this.certChain` etc to null

File: java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java
Comment (lidavidm) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java:252]: Yes, please

Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/client/utils/ClientAuthenticationUtils.java]: This should now be fixed.

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionTlsRootCertsTest.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionTlsRootCertsTest.java]: This should now be done.

File: java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java
Comment (prmoore77) [reply java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/ConnectionMutualTlsTest.java]: This should now be done.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java]: I believe I've addressed these suggestions.  Please confirm.  Thanks.

Comment (prmoore77) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java]: Also - I'm unsure as to Pull Request protocol - should I hit the "Resolve conversation" button when I believe I've addressed the concern, or is that up to the reviewer to do?

**Review by jduo — APPROVED** [review]

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:423]: Filed #38586

**Review by lidavidm — APPROVED** [review]

**Comment by jduo** [comment]
@prmoore77 heads up, after #38521 a copy constructor was added to ArrowFlightSqlClientHandler.Builder
So any new fields that this PR adds will need to be added to that copy constructor.

#38580 also changes ArrowFligthSqlClientHandler to set defaults correctly and adds tests for verifying defaults, so if that goes in first, new fields will be added there too.

**Comment by prmoore77** [comment]
> @prmoore77 heads up, after #38521 a copy constructor was added to ArrowFlightSqlClientHandler.Builder So any new fields that this PR adds will need to be added to that copy constructor.
> 
> #38580 also changes ArrowFligthSqlClientHandler to set defaults correctly and adds tests for verifying defaults, so if that goes in first, new fields will be added there too.

Hi @jduo - I've added the new fields to the new copy constructor (from https://github.com/apache/arrow/pull/38521) - per your recommendation.  Thanks!

**Review by jduo — APPROVED** [review]

**Comment by jbonofre** [comment]
I did a quick test and review. LGTM. Thanks !

**Review by jbonofre — APPROVED** [review]

**Review by lidavidm — APPROVED** [review]

**Pull request merged**
